### PR TITLE
Test invoice payment chargerback reservals

### PR DESCRIPTION
**Pull request by wwjfrodriguez**
I submit this PR for review in that specific endpoint because I need help to understand better how chargerback reservals works. I don't write assertions yet in the test because I'm always recieve a BadRequest:

` KillBillClient::API::BadRequest: {"className":"com.fasterxml.jackson.databind.JsonMappingException","code":null,"message":"Can not instantiate value of type [simple type, class org.killbill.billing.jaxrs.json.InvoicePaymentTransactionJson] from String value ('KillBillClient::Model::InvoicePayment'); no single-String constructor/factory method\n at [Source: org.apache.catalina.connector.CoyoteInputStream@7f233e5; line: 1, column: 1]","causeClassName":null,"causeMessage":null,"stackTrace":[]}
`

Or:

`KillBillClient::API::BadRequest: {"className":"java.lang.IllegalArgumentException","code":null,"message":"transactionExternalKey amount needs to be set","causeClassName":null,"causeMessage":null,"stackTrace":[]}`

### CODE REVIEW

**Comment by pierre** [comment]
I'm guessing these errors come from the fact that the API client isn't correctly implemented, see https://github.com/killbill/killbill-client-ruby/pull/47#discussion_r174361478.

Take a look at this message for instance:

```
KillBillClient::API::BadRequest: {"className":"java.lang.IllegalArgumentException","code":null,"message":"transactionExternalKey amount needs to be set","causeClassName":null,"causeMessage":null,"stackTrace":[]}
```

This message (which has a typo in it unfortunately) comes from: https://github.com/killbill/killbill/blob/6b092a0ca5d812ac6acab938b2b874b59c9abcf9/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/InvoicePaymentResource.java#L251

Some background on chargeback reversals can be found in this ticket: https://github.com/killbill/killbill/issues/477. Here is a test that can provide guidance on how these work: https://github.com/killbill/killbill/blob/6b092a0ca5d812ac6acab938b2b874b59c9abcf9/payment/src/test/java/org/killbill/billing/payment/api/TestPaymentApi.java#L1207.

Basically, the flow is:

1. Create payment (e.g. `PURCHASE`)
2. Trigger chargeback
3. Trigger chargeback reversal

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 1]
```diff
@@ -169,5 +169,40 @@ def test_get_account_invoice_payments
       assert_equal(5.0, account_invoice_payments[1].purchased_amount)
     end
 
+    def test_create_chargeback_and_chargeback_reversal
```
Comment (wwjfrodriguez) [inline killbill-integration-tests/core/test_invoice_payment.rb]: @pierre please verify this.

**Review by wwjfrodriguez — COMMENTED** [review]
@pierre please verify this.

File: killbill-integration-tests/core/test_account.rb [hunk 2]
```diff
@@ -132,54 +132,5 @@ def test_account_update
 
     end
 
-    def test_emails_notifications
```
Comment (pierre) [inline killbill-integration-tests/core/test_account.rb]: ⚠️ It looks like the conflict resolution led to this being deleted.

Might be easier to cherry-pick the good commit (https://git-scm.com/docs/git-cherry-pick) or rebase it (https://git-scm.com/book/en/v2/Git-Branching-Rebasing) on top of a clean branch.

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 3]
```diff
@@ -168,6 +167,40 @@ def test_get_account_invoice_payments
...
+      assert_equal(1, account_transactions.size)
+      assert_equal('PURCHASE', account_transactions[0].transaction_type)
+
+      # Trigger chargerback
+      chargeback = KillBillClient::Model::InvoicePayment.create_chargeback(payment_id, '50.0', 'USD', nil, @user, nil, nil, @options)
+
+      # Verify if a new transaction is created and if their type is CHARGEBACK
+      account_transactions = account.payments(@options).first.transactions
+      assert_equal(2, account_transactions.size)
+      assert_equal('CHARGEBACK', account_transactions[1].transaction_type)
```
Comment (pierre) [inline killbill-integration-tests/core/test_invoice_payment.rb:195]: Can you also verify the status is `SUCCESS`?

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 4]
```diff
@@ -168,6 +167,40 @@ def test_get_account_invoice_payments
...
+
+      # Trigger chargerback reversal
+      transaction_external_key = chargeback.transactions[1].transaction_external_key
+      KillBillClient::Model::InvoicePayment.chargeback_reversal(payment_id, transaction_external_key, nil, @user, nil, nil, @options)
+
+      # Verify if a new transaction is created and if their type is CHARGEBACK
+      account_transactions = account.payments(@options).first.transactions
+      assert_equal(3, account_transactions.size)
+      assert_equal('CHARGEBACK', account_transactions[2].transaction_type)
```
Comment (pierre) [inline killbill-integration-tests/core/test_invoice_payment.rb:206]: Can you also verify the status is `PAYMENT_FAILURE`?

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 5]
```diff
@@ -168,6 +167,40 @@ def test_get_account_invoice_payments
...
+      pay_all_unpaid_invoices(@account.account_id, true, '50.0', @user, @options)
+
+      account = get_account(@account.account_id, true, true, @options)
+      payment_id = account.payments(@options).first.payment_id
+
+      # Verify if a new transaction is created and if their type is PURCHASE
+      account_transactions = account.payments(@options).first.transactions
+      assert_equal(1, account_transactions.size)
+      assert_equal('PURCHASE', account_transactions[0].transaction_type)
+
```
Comment (pierre) [inline killbill-integration-tests/core/test_invoice_payment.rb]: Can you check the account balance here?

```
assert_equal(0, get_account(@account.account_id, true, true, @options).account_balance)
```

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 6]
```diff
@@ -168,6 +167,40 @@ def test_get_account_invoice_payments
...
+      # Trigger chargerback
+      chargeback = KillBillClient::Model::InvoicePayment.create_chargeback(payment_id, '50.0', 'USD', nil, @user, nil, nil, @options)
+
+      # Verify if a new transaction is created and if their type is CHARGEBACK
+      account_transactions = account.payments(@options).first.transactions
+      assert_equal(2, account_transactions.size)
+      assert_equal('CHARGEBACK', account_transactions[1].transaction_type)
+
```
Comment (pierre) [inline killbill-integration-tests/core/test_invoice_payment.rb]: Can you check the account balance here?

```
assert_equal(50, get_account(@account.account_id, true, true, @options).account_balance)
```

File: killbill-integration-tests/core/test_invoice_payment.rb [hunk 7]
```diff
@@ -168,6 +167,40 @@ def test_get_account_invoice_payments
...
+      # Trigger chargerback reversal
+      transaction_external_key = chargeback.transactions[1].transaction_external_key
+      KillBillClient::Model::InvoicePayment.chargeback_reversal(payment_id, transaction_external_key, nil, @user, nil, nil, @options)
+
+      # Verify if a new transaction is created and if their type is CHARGEBACK
+      account_transactions = account.payments(@options).first.transactions
+      assert_equal(3, account_transactions.size)
+      assert_equal('CHARGEBACK', account_transactions[2].transaction_type)
```
Comment (pierre) [inline killbill-integration-tests/core/test_invoice_payment.rb]: Can you check the account balance here?

```
assert_equal(0, get_account(@account.account_id, true, true, @options).account_balance)
```

**Review by pierre — CHANGES_REQUESTED** [review]
Logic looks good, let's just make a few more assumptions around the account balance.

This is to make sure the right endpoints are called (if we were to call the pure payment APIs instead of the invoice payment APIs, the account balance wouldn't be updated).

**Comment by wwjfrodriguez** [comment]
@pierre I already implemented all your requested changes, please verify.

**Review by pierre — APPROVED** [review]
